BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TransmissionProcessor.cs

```diff
@@ -29,6 +30,11 @@ public void Process(ITelemetry item)
         {
             TelemetryDebugWriter.WriteTelemetry(item);
 
+            if (string.IsNullOrWhiteSpace(item.Context.InstrumentationKey))
+            {
+                CoreEventSource.Log.TransmissionProcessorNoInstrumentationKey();
```

Comment:
Just to keep in mind - some custom channel implementations maybe OK with no instrumentation key, such as custom file writer to save telemetry locally.

Also, AI VS Output / AI VS Integration UX may not need IKey to power up its experiences - will we emit this message for those scenarios? These scenarios might be fairly common, we can put #if !debug here to avoid this message in VS/Debug.

Reply:
The `#if !debug` would be a compile time switch. I don't think we can detect when a consumer is in debug mode

Reply:
Even if we do it, we should Log it once maybe, rather than for every item? Or maybe once in few seconds..

Reply:
@Dmitry-Matveev, I don't know how to balance the concern that an empty ikey could be valid. I don't know enough about these scenarios. 
Do you think it's valid to optimize for the most common/expected scenarios?

@cijothomas, I agree for every item is too much, but I'm reluctant to log only once because I don't want the log to get missed. I can ask Raj for a second opinion on how frequent he thinks the log should be to get recognized.

Reply:
Something like `Debugger.IsAttached` can be a better pick here - does not execute locally in VS, does execute when deployed to PROD. One notable issue - log won't be published is you debug in production, e.g. with WinDbg or remote debugging, but that should not matter already since you are capturing everything you need in debugger.

Reply:
@Dmitry-Matveev's comment about channels got me thinking about what if we only set this on our 1st party channels.
And I just discovered the log I'm adding already exists on both!
However, both channels restrict this to Verbose logs only. 
I'm not sure if our troubleshooting guides are collecting this. I will investigate...

https://github.com/microsoft/ApplicationInsights-dotnet/blob/7633ae849edc826a8547745b6bf9f3174715d4bd/BASE/src/Microsoft.ApplicationInsights/Channel/InMemoryChannel.cs#L136-L144

https://github.com/microsoft/ApplicationInsights-dotnet/blob/4ff3b7fb22274321671058e47a65738c61e78438/BASE/src/ServerTelemetryChannel/ServerTelemetryChannel.cs#L266-L274